### ISSUE TITLE
Assert failure in ImageCache

### ISSUE THREAD

**Issue by HughMacdonald** [OpenImageIO/oiio#1239]
I've just found myself getting an assertion failure in ImageCache.

ImageCache should be thread-safe, right? I'm only getting this error when I run the processing in multiple threads in parallel.

Here's the error:

    /Users/john/dev/gafferDependencies.git/oiio-Release-1.5.17/src/libtexture/imagecache.cpp:2097:
        failed assertion 'sweep != m_tilecache.end() && "no way m_tilecache can be empty and use
        too much memory"'

I'm doing this through Gaffer, but what I'm doing is:

* Parallel loading 64x64 tiles from a 4096x3112 image
* Clear the ImageCache
* Repeat another 19 times

It seems as though it mainly happens when I have more than 10 repeats, but I have seen it happen with fewer. I'm guessing it's just chance. It's also more common the more threads I have processing.

Any ideas?

**Comment by HughMacdonald** [comment]
Ah yes - this is on OSX, too (10.10.5). I'm running a build and a test on Linux at the moment, so I'll let you know how that one turns out.

**Comment by HughMacdonald** [comment]
Just to confirm - I am seeing the same error on Linux (Ubuntu 12.04)

**Comment by lgritz** [comment]
IC is certainly threadsafe.

Is the image file itself tiled? How big have you set the IC tile cache? ("max_memory_MB")

I can imagine situations in which it's the assertion itself that is incorrect.

**Comment by lgritz** [comment]
Aha, I think I know what's going on. When you "clear the image cache", are you doing an invalidate()?

**Comment by HughMacdonald** [comment]
I had thought it must be!

The image I am loading is a scanline EXR (a checkerboard rendered out of Nuke)

I'm just looking at the code that's calling ImageCache (I wasn't involved with the current code myself), and it doesn't look as though the cache size is being set. What's the default size?

The cache is being cleared with:

    cache->invalidate_all( true );

**Comment by lgritz** [comment]
Some important background: Each thread retains a reference-counted pointer to the last two tiles it's accessed, so reusing the same texture as one of the last two queries doesn't even need to search the main cache, and also so that even if those tiles are kicked out of the main cache, they won't disappear from the thread's view mid-texture-access. 

So.. Here's a sequence that I think could trigger the assertion:

1. A bunch of threads access various tiles, thus setting their two texture texture tile references.

2. The cache is cleared/invalidated, but that doesn't free the tiles held by threads immediately (because those threads may literally be accessing those tiles), but just marks the per-thread record as "purge your microcache next time you have a chance and it's safe."

3. If the held tiles are a disjoint set and you set the max_memory_MB low enough, it's possible that those held tiles could exceed your prescribed cache size.

4. So next tile you load will want to free something from the main cache to get you below the limit. It will sweep through the whole list, nothing left but still not below the limit. Thus tripping the assertion.

It's harmless. These are "dead men walking" tiles that will be freed as soon each thread is asked to retrieve a tile (at which point, if its "purge" flag is set, it will free what it's holding). But momentarily, it seems to be in the impossible situation of being over its memory limit but having nothing ready to free.

**Comment by lgritz** [comment]
Yeah! If it's a scanline exr, that means that the whole image is one huge tile, which just might exceed your cache size. So it frees the tile from the main cache, but your're still over the memory limit.

Harmless, but tricky.

**Comment by lgritz** [comment]
I have a solution that I discussed in another thread:

I'm thinking about changing the logic so that the max_memory_MB is forced to be increased any time a single tile is encountered that exceeds it. So, it will make it possible (via poor choice of cache size for the kinds of files you're using) to exceed the memory bounds you requested. But it will make it impossible to set it the cache limit so low that you get seemingly impossible situations (like simultaneously exceeding your budget yet having nothing able to be freed).

**Comment by HughMacdonald** [comment]
Ahh - that would make sense then!

Weirdly, I only came across this because I was doing some performance-testing of some new code I've been writing for parallel processing. My new code doesn't cause this error at all - it was only when I was comparing the same test on the old code that it came up.

The main difference is that there was a bug in the old code that caused the same tile (in Gaffer's world) and channel to be requested twice in quick succession on different threads.

Is the solution then to lose the assert() call? Or is there a more thorough check that could be done there instead?


Thanks for the explanation!

**Comment by HughMacdonald** [comment]
Ah yes - that would make sense.

I think that we definitely need to look into setting the cache size more accurately ourselves - I'll have to have a chat with @johnhaddon about how that would be best to do. That in addition to the change discussed elsewhere would certainly solve the problems here.


Thanks

**Comment by lgritz** [comment]
We could change it to a DASSERT. That would catch a true leak in a debug build, but still allow this to be fine for a production build.

**Comment by lgritz** [comment]
Or get rid of the assertion entirely. It was designed to catch bugs. At this point (dozens of feature films), I'm fairly confident that the algorithm is sound, so the assertion is probably unnecessary.

**Comment by lgritz** [comment]
OK, how about this:

1. I'll immediately submit a PR changing most of the ASSERT to DASSERT, this code has been stable for a long time. That should get you unstuck.

2. I'll add to my list the task of bumping up cache size automatically when too-big tiles are found, so we never have the situation of tiles bigger than cache budget. This will be necessary for the deep-IC anyway.

**Comment by HughMacdonald** [comment]
That sounds great.

It's not an error we'd come across before, and then @andrewkaufmann and I hit upon it at about the same time. We ought to be setting our cache to be larger anyway, which may well get us around this.

Thanks for the speedy response on this!

**Comment by lgritz** [comment]
Proposed temporary fix in #1240.

More robust long-term fix will be part of the anticipated near-future support of deep data in ImageCache.

**Comment by MrKepzie** [comment]
@lgritz There seem to be a remaining bug in the cache:
If you pass chbegin/chend (e.g = 0 / 4) but cache_chbegin=0 and cache_chend=-1, the resulting image is  wrongly filled, assuming that the buffer is large enough to store only channels 0-3. You can try it out on one of the EXR I sent you with the image sequence of the unit test.

**Comment by lgritz** [comment]
Oops, sorry, I'll take a look today.

**Comment by MrKepzie** [comment]
@lgritz I'm hitting another bug as well:

If  a record of the last file read is kept and this file is invalidated if a different file is read, then we may enter a race condition with multiple threads:

Thread A:
-----------------
- *ImageCacheImpl::invalidate(filename)* will call *ImageCacheFile::invalidate()* which in turns calls *invalidate_spec()*  which clears *m_subimages*. This operation is **protected** by *m_input_mutex*.
- Set the purge = 1 atomic int  in *purge_perthread_microcaches()*, actually *queueing* a purge of the *tile* pointer on **all** the per-threads data (meaning thread B included)

Thread B:
-------------
- Inside ImageCacheImpl::get_pixels, before *purge_perthread_microcaches()* is called from thread A, so the purge is not actually done by thread B yet (i.e: the *tile* pointer is not reset)
- The *tile* member on the per-thread data of thread B is in a dirty state: its spec has been invalidated but the pointer is still around
- An assertion is triggered right after in get_pixels when entering this line:

https://github.com/OpenImageIO/oiio/blob/master/src/libtexture/imagecache.cpp#L2680
Which in turns calls
https://github.com/OpenImageIO/oiio/blob/master/src/libtexture/imagecache.cpp#L1389
Which leads to this assert being triggered:
https://github.com/OpenImageIO/oiio/blob/master/src/libtexture/imagecache_pvt.h#L265

It is triggered because the spec is invalidated (so m_subimages has been cleared) but the pointer can still be accessed.

**Comment by lgritz** [comment]
For the get_pixel issue, you are right, and my fix is in #1246

**Comment by MrKepzie** [comment]
Will check.
Btw on the race-condition, do  you want me to open-up a new issue?
It appears to me that there are 2 possible solutions:

- Either take a global lock (ruining multi-threading)
- Let the last thread using the tile (i.e: when the use_count of the ref counted ptr drops to 1)  call invalidate on the ImageCacheFile instead of doing it immediately whilst others could still be using that tile.

**Comment by lgritz** [comment]
Yeah, I think we should move the race condition to a new issue.
I'm still mulling over that problem.
